File: app/android/app/src/main/kotlin/com/gymbuddy/gymbuddy/glasses/GlassesChannel.kt
package com.gymbuddy.gymbuddy.glasses

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * The Android side of the glasses platform channel — the native counterpart of
 * the Dart `MetaGlassesSensorSource` (M7). It exposes a single control
 * [MethodChannel] plus two streaming [EventChannel]s (reps + form cues), all
 * backed by a [DatSdkClient] that wraps the Meta DAT SDK's camera/audio/mic.
 *
 * Wire contract (kept identical on iOS and in the Dart binding):
 *  - method channel `gymbuddy/glasses`:
 *      `connect`       -> { availability: "available"|"unavailable",
 *                           capabilities: { repCounting: Bool, formTracking: Bool } }
 *      `startTracking` (args { name: String, formTracked: Bool }) -> null
 *      `stopTracking`  -> null
 *      `playCue`       (args { message: String }) -> null
 *      `dispose`       -> null
 *  - event channel `gymbuddy/glasses/reps`     -> { index: Int, timestampMs: Long, confidence: Double? }
 *  - event channel `gymbuddy/glasses/formCues` -> { severity: String, message: String }
 *
 * Because the DAT SDK isn't bundled in a no-hardware build, [connect] resolves to
 * `unavailable` and no events are emitted — the Dart layer then falls back to the
 * mock, so the channel is always safe to wire up regardless of hardware.
 */
class GlassesChannel(
    context: Context,
    messenger: BinaryMessenger,
) : MethodChannel.MethodCallHandler {

    private val client: DatSdkClient = DatSdkClient.create(context.applicationContext)
    private val mainHandler = Handler(Looper.getMainLooper())

    private val methodChannel = MethodChannel(messenger, METHOD_CHANNEL).apply {
        setMethodCallHandler(this@GlassesChannel)
    }

    private var repsSink: EventChannel.EventSink? = null
    private val repsChannel = EventChannel(messenger, REPS_CHANNEL).apply {
        setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                repsSink = events
            }

            override fun onCancel(arguments: Any?) {
                repsSink = null
            }
        })
    }

    private var formCuesSink: EventChannel.EventSink? = null
    private val formCuesChannel = EventChannel(messenger, FORM_CUES_CHANNEL).apply {
        setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                formCuesSink = events
            }

            override fun onCancel(arguments: Any?) {
                formCuesSink = null
            }
        })
    }

    /** Forwards DAT SDK events onto the Flutter sinks; sink calls must hit the main thread. */
    private val trackingListener = object : TrackingListener {
        override fun onRep(sample: RepSample) {
            mainHandler.post {
                repsSink?.success(
                    mapOf(
                        "index" to sample.index,
                        "timestampMs" to sample.timestampMs,
                        "confidence" to sample.confidence,
                    ),
                )
            }
        }

        override fun onFormCue(cue: FormCueSample) {
            mainHandler.post {
                formCuesSink?.success(
                    mapOf(
                        "severity" to cue.severity,
                        "message" to cue.message,
                    ),
                )
            }
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "connect" -> {
                val availability = client.connect()
                val caps = client.capabilities()
                result.success(
                    mapOf(
                        "availability" to availability.name.lowercase(),
                        "capabilities" to mapOf(
                            "repCounting" to caps.repCounting,
                            "formTracking" to caps.formTracking,
                        ),
                    ),
                )
            }

            "startTracking" -> {
                val name = call.argument<String>("name")
                if (name == null) {
                    result.error("bad_args", "startTracking requires a 'name'", null)
                    return
                }
                val formTracked = call.argument<Boolean>("formTracked") ?: false
                client.startTracking(TrackedExercise(name, formTracked), trackingListener)
                result.success(null)
            }

            "stopTracking" -> {
                client.stopTracking()
                result.success(null)
            }

            "playCue" -> {
                val message = call.argument<String>("message")
                if (message == null) {
                    result.error("bad_args", "playCue requires a 'message'", null)
                    return
                }
                client.playCue(message)
                result.success(null)
            }

            "dispose" -> {
                client.dispose()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    /** Tears down all channels and the underlying client; call from the activity's engine teardown. */
    fun dispose() {
        methodChannel.setMethodCallHandler(null)
        repsChannel.setStreamHandler(null)
        formCuesChannel.setStreamHandler(null)
        repsSink = null
        formCuesSink = null
        client.dispose()
    }

    companion object {
        const val METHOD_CHANNEL = "gymbuddy/glasses"
        const val REPS_CHANNEL = "gymbuddy/glasses/reps"
        const val FORM_CUES_CHANNEL = "gymbuddy/glasses/formCues"
    }
}

File: app/ios/Runner/Glasses/GlassesChannel.swift
import Flutter
import Foundation

/// The iOS side of the glasses platform channel — the native counterpart of the
/// Dart `MetaGlassesSensorSource` (M7) and the mirror of the Android
/// `GlassesChannel`. It exposes a single control `FlutterMethodChannel` plus two
/// streaming `FlutterEventChannel`s (reps + form cues), all backed by a
/// `DatSdkClient` that wraps the Meta DAT SDK's camera/audio/mic.
///
/// Wire contract (kept identical on Android and in the Dart binding):
///  - method channel `gymbuddy/glasses`:
///      `connect`       -> { availability: "available"|"unavailable",
///                           capabilities: { repCounting: Bool, formTracking: Bool } }
///      `startTracking` (args { name: String, formTracked: Bool }) -> null
///      `stopTracking`  -> null
///      `playCue`       (args { message: String }) -> null
///      `dispose`       -> null
///  - event channel `gymbuddy/glasses/reps`     -> { index: Int, timestampMs: Int, confidence: Double? }
///  - event channel `gymbuddy/glasses/formCues` -> { severity: String, message: String }
///  - event channel `gymbuddy/glasses/wake`     -> { timestampMs: Int, phrase: String, confidence: Double? }
///
/// Because the DAT SDK isn't bundled in a no-hardware build, `connect` resolves to
/// `unavailable` and no events are emitted — the Dart layer then falls back to the
/// mock, so the channel is always safe to wire up regardless of hardware.
final class GlassesChannel: NSObject {
  static let methodChannelName = "gymbuddy/glasses"
  static let repsChannelName = "gymbuddy/glasses/reps"
  static let formCuesChannelName = "gymbuddy/glasses/formCues"
  static let wakeChannelName = "gymbuddy/glasses/wake"

  private let client: DatSdkClient
  private let methodChannel: FlutterMethodChannel
  private let repsChannel: FlutterEventChannel
  private let formCuesChannel: FlutterEventChannel
  private let wakeChannel: FlutterEventChannel

  private let repsStreamHandler = QueuingStreamHandler()
  private let formCuesStreamHandler = QueuingStreamHandler()
  // The wake channel is its own stream because wake detection is always-on and
  // independent of tracking; the listener is armed only while Flutter is
  // listening, and torn down on cancel.
  private var wakeStreamHandler: QueuingStreamHandler!

  init(messenger: FlutterBinaryMessenger) {
    self.client = DatSdkClientFactory.create()
    self.methodChannel = FlutterMethodChannel(
      name: GlassesChannel.methodChannelName, binaryMessenger: messenger)
    self.repsChannel = FlutterEventChannel(
      name: GlassesChannel.repsChannelName, binaryMessenger: messenger)
    self.formCuesChannel = FlutterEventChannel(
      name: GlassesChannel.formCuesChannelName, binaryMessenger: messenger)
    self.wakeChannel = FlutterEventChannel(
      name: GlassesChannel.wakeChannelName, binaryMessenger: messenger)
    super.init()

    methodChannel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }
    repsChannel.setStreamHandler(repsStreamHandler)
    formCuesChannel.setStreamHandler(formCuesStreamHandler)
    wakeStreamHandler = QueuingStreamHandler(
      onListen: { [weak self] in self.map { $0.client.setWakeListener($0) } },
      onCancel: { [weak self] in self?.client.setWakeListener(nil) })
    wakeChannel.setStreamHandler(wakeStreamHandler)
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "connect":
      let availability = client.connect()
      let caps = client.capabilities()
      result([
        "availability": availability.rawValue,
        "capabilities": [
          "repCounting": caps.repCounting,
          "formTracking": caps.formTracking,
        ],
      ])

    case "startTracking":
      let args = call.arguments as? [String: Any]
      guard let name = args?["name"] as? String else {
        result(
          FlutterError(
            code: "bad_args", message: "startTracking requires a 'name'", details: nil))
        return
      }
      let formTracked = args?["formTracked"] as? Bool ?? false
      client.startTracking(
        TrackedExercise(name: name, formTracked: formTracked), listener: self)
      result(nil)

    case "stopTracking":
      client.stopTracking()
      result(nil)

    case "playCue":
      let args = call.arguments as? [String: Any]
      guard let message = args?["message"] as? String else {
        result(
          FlutterError(
            code: "bad_args", message: "playCue requires a 'message'", details: nil))
        return
      }
      client.playCue(message)
      result(nil)

    case "dispose":
      client.dispose()
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Tears down all channels and the underlying client; call from the app's engine teardown.
  func dispose() {
    methodChannel.setMethodCallHandler(nil)
    repsChannel.setStreamHandler(nil)
    formCuesChannel.setStreamHandler(nil)
    wakeChannel.setStreamHandler(nil)
    client.setWakeListener(nil)
    client.dispose()
  }
}

// MARK: - TrackingListener

extension GlassesChannel: TrackingListener {
  func onRep(_ sample: RepSample) {
    // Flutter event sinks must be invoked on the platform (main) thread.
    DispatchQueue.main.async { [weak self] in
      self?.repsStreamHandler.send([
        "index": sample.index,
        "timestampMs": sample.timestampMs,
        "confidence": sample.confidence as Any,
      ])
    }
  }

  func onFormCue(_ cue: FormCueSample) {
    DispatchQueue.main.async { [weak self] in
      self?.formCuesStreamHandler.send([
        "severity": cue.severity,
        "message": cue.message,
      ])
    }
  }
}

// MARK: - WakeListener

extension GlassesChannel: WakeListener {
  func onWake(_ sample: WakeSample) {
    DispatchQueue.main.async { [weak self] in
      self?.wakeStreamHandler.send([
        "timestampMs": sample.timestampMs,
        "phrase": sample.phrase,
        "confidence": sample.confidence as Any,
      ])
    }
  }
}

// MARK: - QueuingStreamHandler

/// A minimal `FlutterStreamHandler` that holds the active sink between
/// `onListen`/`onCancel`, mirroring the Android channel's nullable sink fields.
/// Emissions before a listener attaches (or after it cancels) are dropped — the
/// same posture as the Android `EventSink?`.
final class QueuingStreamHandler: NSObject, FlutterStreamHandler {
  private var sink: FlutterEventSink?
  private let onListenCallback: (() -> Void)?
  private let onCancelCallback: (() -> Void)?

  /// `onListen`/`onCancel` fire as a Flutter listener attaches/detaches — used by
  /// the wake channel to arm and tear down the always-on mic only while someone
  /// is listening.
  init(onListen: (() -> Void)? = nil, onCancel: (() -> Void)? = nil) {
    self.onListenCallback = onListen
    self.onCancelCallback = onCancel
  }

  func send(_ event: Any) {
    sink?(event)
  }

  func onListen(
    withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    sink = events
    onListenCallback?()
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    sink = nil
    onCancelCallback?()
    return nil
  }
}
